text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,328 @@
+
+A `cargo build` in `project-1` will produce new artifacts in `/cargo-cache/project-1-<hash>/debug/...`.
+
+A `cargo clean` will only remove the `/cargo-cache/<project>/` subdirectory, not all the artifacts.
+
+In this situation, it's not possible for Cargo to produce invalid state without a `build.rs` deliberately writing outside its target directory.
+
+Two projects can be built in parallel without troubles.
+
+#### With both set
+
+`CARGO_TARGET_DIR` was present long before `CARGO_TARGET_DIRECTORIES`: backward compatibility is important, so the first always trumps the second,
+there is no mixing going on.
+
+#### Absolute and relative paths
+
+`CARGO_TARGET_DIR` can be either a relative or absolute path, which makes sense since it's mostly intended for a single project, which can then
+work from its own position to configure the target directory.
+
+On the other hand `CARGO_TARGET_DIRECTORIES` is intended to be used with several projects, possibly completely unrelated to each other. As such,
+it does not accept relative paths, only absolute ones. If a compelling use case is present for a relative path, it can added in the future as a
+backward-compatible change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Setting `CARGO_TARGET_DIRECTORIES`
+
+The option is similar to `CARGO_TARGET_DIR` and can be set in the same places. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-directories = "/absolute/path/to/target/directories"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIRECTORIES="/absolute/path/to/target/directories" cargo build`
+- Through the command line flag: `cargo build --target-directories /absolute/path/to/target/directories`
+
+The given path must be absolute: setting `CARGO_TARGET_DIRECTORIES` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).
+
+## Resolution order relative to `CARGO_TARGET_DIR`
+
+The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:
+
+`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-directories` > `CARGO_TARGET_DIRECTORIES` > `target-directories = ...`
+
+## Naming
+
+In the example in the previous section, using `CARGO_TARGET_DIRECTORIES` with `cargo build` produces named subdirectories. The name of those is partially deterministic:
```

Comment:
Under bazel, you mention the hash is after resolving symlinks.  Should we call that out here as well?  Should we resolve symlinks before extracting the parent directory name?

Reply:
`cargo metadata` currently resolves symlinks, we should just keep the existing behaviour

Reply:
I don't see how `cargo metadata` is relevant; we are talking about the behavior of the code that generates what `cargo metadata` produces and we need to document in whether symlinks are resolved before hashing.  I see this being a clarification within the naming section on what we are hashing.

Reply:
done